src/libraries/System.Private.CoreLib/src/System/Buffers/ArrayPool.cs

```diff
@@ -18,9 +18,9 @@ namespace System.Buffers
     /// </remarks>
     public abstract class ArrayPool<T>
     {
-        // Store the shared ArrayPool in a field of its derived sealed type so the Jit can "see" the exact type
-        // when the Shared property is inlined which will allow it to devirtualize calls made on it.
-        private static readonly SharedArrayPool<T> s_shared = new SharedArrayPool<T>();
+        private static readonly ArrayPool<T> s_shared = SharedArrayPoolStatics.s_diagnosticPoolEnabled ?
```

Comment:
The diagnostic pool in this PR is expensive, but we could possibly have this turned on automatically when in debug / optimizations disabled, at least for coreclr.